### PR TITLE
os/bluestore: space efficient int encoding

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/include/small_encoding.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// vim: ts=8 sw=2 smarttab
+#ifndef CEPH_SMALL_ENCODING_H
+#define CEPH_SMALL_ENCODING_H
+
+#include "include/buffer.h"
+#include "include/int_types.h"
+
+// first(low) 2 bits = how many low nibbles are zeros
+// high bit of each byte = another byte follows
+// (so, 5 bits data in first byte, 7 bits data thereafter)
```
Comment (liewegas) [inline src/include/small_encoding.h]: Could also make 00 = no low zeros, 01 = 6 low zeros, 02 = 10 low zeros, 03 = 14 low zeros, or something like that if we think the <6 or even <12 low zero bits is less common to optimize for.  Remember block_size is 4096 = 12 bits.

**Comment by jjhuo** [comment]
rocksdb implementation has Varint64 as well.
ceph/src/rocksdb/util/coding.h

File: src/include/small_encoding.h [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+  while (byte & 0x80) {
+    ::decode(byte, p);
+    v |= (T)(byte & 0x7f) << shift;
+    shift += 7;
+  }
+}
+
+// signed varint + lowz encoding
+//
+// first(low) 2 bits = how many low zero bits (nibbles)
```
Comment (ifed01) [inline src/include/small_encoding.h]: lacks description for the first bit - sign

File: src/include/small_encoding.h [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    shift += 7;
+  }
+  if (negative) {
+    v = -v;
+  }
+}
+
+
+template<typename T>
+inline void small_delta_encode_varint_lowz(T prev, uint32_t cur,
```
Comment (ifed01) [inline src/include/small_encoding.h]: Lacks corresponding decode. No much value in this method though, can be easily implemented via small_encode_variant_lowz...

File: src/include/small_encoding.h [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+					   bufferlist& bl) {
+  int32_t v = cur - prev;
+  uint8_t byte = 0;
+  if (v < 0) {
+    byte |= 1;
+    v = -v;
+  }
+  int lowz = v ? std::min((int)ctz(v) - 8, 0) : 0;
```
Comment (ifed01) [inline src/include/small_encoding.h]: "- 8" -> "/4"

File: src/include/small_encoding.h [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+  word |= (v << pos) & 0x7fffffff;
+  v >>= 31 - pos;
+  if (!v) {
+    ::encode(word, bl);
+    return;
+  }
+  word |= 0x80000000;
+  ::encode(word, bl);
+  uint8_t byte = v & 0x7f;
+  v >>= 7;
```
Comment (ifed01) [inline src/include/small_encoding.h]: use small_encode_varint for the rest?

File: src/include/small_encoding.h [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+  }
+  word |= (v << pos) & 0x7fffffff;
+  v >>= 31 - pos;
+  if (!v) {
+    ::encode(word, bl);
+    return;
+  }
+  word |= 0x80000000;
+  ::encode(word, bl);
+  uint8_t byte = v & 0x7f;
```
Comment (ifed01) [inline src/include/small_encoding.h]: use small_encode_varint for the rest?

File: src/include/small_encoding.h [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+  while (v) {
+    byte |= 0x80;
+    ::encode(byte, bl);
+    byte = (v & 0x7f);
+    v >>= 7;
+  }
+  ::encode(byte, bl);
+}
+
+inline void small_delta_decode_lba(uint64_t pevious, uint64_t& v,
```
Comment (ifed01) [inline src/include/small_encoding.h]: implementation?

File: src/include/small_encoding.h [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+//
+// first (low) bit = sign (0 = positive, 1 = negative)
+// next 2 bits = how many low zero bits
+//   00 = none
+//   01 = 12
+//   10 = 16
+//   11 = 20
+// high bit of each byte = another byte follows
+// (so, 4 bits data in first byte, 7 bits data thereafter)
+inline void small_delta_encode_lba(uint64_t previous, uint64_t lba,
```
Comment (ifed01) [inline src/include/small_encoding.h]: Really needed?
Implement via small_encode_lba?

File: src/include/small_encoding.h [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+  bl.append(bp);
+}
+template<typename T>
+inline void small_decode_buf_lowz(T& bp, bufferlist::iterator& p) {
+  size_t l;
+  small_decode_varint_lowz(l, p);
+  p.copy(l, bp);
+}
+
+// STL containters
```
Comment (ifed01) [inline src/include/small_encoding.h]: typo

File: src/os/bluestore/bluestore_types.cc [hunk 10]
```diff
@@ -98,6 +98,44 @@ void bluestore_cnode_t::generate_test_instances(list<bluestore_cnode_t*>& o)
...
+void small_encode(const vector<bluestore_pextent_t>& v, bufferlist& bl)
+{
+  size_t n = v.size();
+  small_encode_varint(n, bl);
+  if (n) {
+    auto p = v.cbegin();
+    p->encode(bl);
+    int64_t pos = p->offset;
+    for (++p; p != v.cend(); ++p) {
+      small_encode_signed_varint_lowz((int64_t)p->offset - pos, bl);
```
Comment (ifed01) [inline src/os/bluestore/bluestore_types.cc]: Probably no much sence in encoding delta here

File: src/os/bluestore/bluestore_types.cc [hunk 11]
```diff
@@ -286,16 +324,23 @@ bool bluestore_extent_ref_map_t::intersects(
 
 void bluestore_extent_ref_map_t::encode(bufferlist& bl) const
 {
-  ENCODE_START(1, 1, bl);
-  ::encode(ref_map, bl);
-  ENCODE_FINISH(bl);
+  uint32_t n = ref_map.size();
+  small_encode_varint(n, bl);
+  for (auto& p : ref_map) {
+    small_encode_varint_lowz(p.first, bl);
```
Comment (ifed01) [inline src/os/bluestore/bluestore_types.cc]: and here delta version is probably more beneficial...

**Comment by ifed01** [comment]
LGTM

**Pull request merged**